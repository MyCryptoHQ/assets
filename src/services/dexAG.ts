import { Asset } from '../constants';
import { CachedResponse, getInitialCache, withCache } from './cache';
import { isSimilar } from '../utils';

const DEXAG_API_ENDPOINT = 'https://api.dex.ag/token-list-full';

interface DexAgCoin {
  symbol: string;
  name: string;
}

interface DexAgData {
  coins: DexAgCoin[];
}

/**
 * Match an asset with a DexAg ID.
 *
 * @param {Partial<DexAgData>} cache
 * @param {Asset} asset
 * @return {Promise<CachedResponse<DexAgData, string | null>>}
 */
export const matchDexAgId = async (
  cache: Partial<DexAgData>,
  asset: Asset
): Promise<CachedResponse<DexAgData, string | null>> => {
  const data = cache;
  if (!data || !data.coins) {
    data.coins = await getInitialCache<DexAgCoin[]>(DEXAG_API_ENDPOINT);
  }

  const match = data.coins.find(coin => coin.symbol.toLowerCase() === asset.symbol.toLowerCase());
  if (match) {
    return {
      cache: data,
      data: match.symbol
    };
  }

  return {
    cache: data,
    data: null
  };
};

export default withCache<DexAgData, typeof matchDexAgId, string | null>(matchDexAgId);
