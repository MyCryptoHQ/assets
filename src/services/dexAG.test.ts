import { matchDexAgId } from './dexAG';
import { matchDexAgId as matchDexAgIdWithCache } from './index';
import { getInitialCache } from './cache';

const CACHE_DATA = [{ symbol: 'ETH', name: 'ETH' }];

jest.mock('./cache', () => ({
  getInitialCache: jest.fn(async () => CACHE_DATA),

  withCache: jest.requireActual('./cache').withCache
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('matchDexAgId', () => {
  it('initializes the cache', async () => {
    const asset = {
      name: 'ETH',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    const { cache } = await matchDexAgId({}, asset);
    const result = await matchDexAgId(cache, asset);

    expect(result.cache.coins).toMatchSnapshot();
    expect(getInitialCache).toHaveBeenCalledTimes(1);
  });

  it('fetches the ID of an asset', async () => {
    const asset = {
      name: 'ETH',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    await expect(matchDexAgId({}, asset)).resolves.toEqual({
      cache: {
        coins: CACHE_DATA
      },
      data: 'ETH'
    });
  });

  it('returns null if an asset could not be found', async () => {
    const asset = {
      name: 'Ethereum Classic',
      symbol: 'ETC',
      uuid: '6c1e671f-9af5-546d-9c1a-52067bdf0e53'
    };

    await expect(matchDexAgId({}, asset)).resolves.toEqual({
      cache: {
        coins: CACHE_DATA
      },
      data: null
    });
  });

  it('works with a cache', async () => {
    const asset = {
      name: 'ETH',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    await expect(matchDexAgIdWithCache(asset)).resolves.toBe('ETH');
  });
});
