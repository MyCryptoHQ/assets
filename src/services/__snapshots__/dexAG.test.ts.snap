// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchDexAgId initializes the cache 1`] = `
Array [
  Object {
    "name": "ETH",
    "symbol": "ETH",
  },
]
`;
